src/serac/physics/utilities/functional/domain_integral_cuda.cuh

```diff
@@ -0,0 +1,309 @@
+#pragma once
+
+#include "mfem.hpp"
+#include "mfem/linalg/dtensor.hpp"
+
+#include "serac/physics/utilities/functional/integral_utilities.hpp"
+#include "serac/physics/utilities/functional/domain_integral_shared.hpp"
+#include <cstring>
+
+namespace serac {
+
+namespace detail {
+template <typename T, typename... Dims>
+__host__ inline mfem::DeviceTensor<sizeof...(Dims),T> Reshape(T *ptr, Dims... dims)
+   {
+      return mfem::DeviceTensor<sizeof...(Dims),T>(ptr, dims...);
+   }
+
+
+  /**
+   * @brief utility method to display last cuda error message
+   *
+   * @param[in] o The output stream to post success or CUDA error messages
+   * @param[in] success_string A string to print if there are no CUDA error messages
+   */
+  inline void displayLastCUDAErrorMessage(std::ostream & o, const char * success_string = "") {
+      auto error = cudaGetLastError();
+      if(error != cudaError::cudaSuccess) {
+	o << "Last CUDA Error Message :" << cudaGetErrorString(error) << std::endl;
+      } else if (strlen(success_string) > 0) {
+	o << success_string << std::endl;
+      }
+    }
+
+} // namespace detail
+
+  namespace domain_integral {  
+
+  template <Geometry g, typename test, typename trial, int Q, typename derivatives_type, typename lambda, typename u_type, typename r_type, typename J_type, typename X_type>
+  __global__ void eval_cuda_element(const u_type u, r_type r, derivatives_type * derivatives_ptr, J_type J, X_type X, int num_elements, lambda qf) {
+
+    using test_element               = finite_element<g, test>;
+    using trial_element              = finite_element<g, trial>;
+    using element_residual_type      = typename trial_element::residual_type;
+    static constexpr auto rule       = GaussQuadratureRule<g, Q>();
+
+    // for each element in the domain
+    const int grid_stride = blockDim.x * gridDim.x;
+#pragma unroll
```

Comment:
does clang-format flush this all the way to the left? I usually prefer if `#pragma unroll` is aligned with its corresponding for loop